server/s3_list_object_versions_action.cc

```diff
@@ -102,20 +100,26 @@ void S3ListObjectVersionsAction::validate_request() {
   // Validate the input max-keys
   std::string max_k = request->get_query_string_value("max-keys");
   if (max_k.empty()) {
-    max_keys = 1000;
+    max_keys = kMaxKeys;
   } else {
     int max_keys_temp = 0;
-    if ((!S3CommonUtilities::stoi(max_k, max_keys_temp)) ||
-        (max_keys_temp < 0)) {
+    if (!S3CommonUtilities::stoi(max_k, max_keys_temp)) {
+      s3_log(S3_LOG_DEBUG, request_id, "invalid max-keys = %s\n",
+             max_k.c_str());
+      set_s3_error("InvalidArgument");
+      set_s3_error_message(
+          "Provided max-keys not an integer or within integer range");
+      set_invalid_argument("max-keys", max_k);
+      send_response = true;
+    } else if (max_keys_temp < 0) {
       s3_log(S3_LOG_DEBUG, request_id, "invalid max-keys = %s\n",
              max_k.c_str());
-      // TODO: Add invalid argument details to the response.
       set_s3_error("InvalidArgument");
       set_s3_error_message("max-keys cannot be negative");
-      set_invalid_argument("max-keys", max_k.c_str());
+      set_invalid_argument("max-keys", max_k);
       send_response = true;
     } else {
-      max_keys = max_keys_temp;
+      max_keys = max_keys_temp > kMaxKeys ? kMaxKeys : max_keys_temp;
```

Comment:
For future reference, you can just use [std::min](https://en.cppreference.com/w/cpp/algorithm/min).

```suggestion
      max_keys = std::min(max_keys_temp, kMaxKeys);
```